text/0000-let-else.md

```diff
@@ -0,0 +1,520 @@
+accessible as they would normally be.
+
+# Drawbacks
+[drawbacks]: #drawbacks
+
+## The diverging block
+
+"Must diverge" is an unusual requirement, which doesn't exist elsewhere in the language as of the time of writing, 
+and might be difficult to explain or lead to confusing errors for programmers new to this feature.
+
+This also necessitates a new block expression subtype, something like `BlockExpressionDiverging`.
+
+## `let PATTERN = if {} else {} else {};`
+
+One unfortunate combination of this feature with regular if-else expressions is the possibility of `let PATTERN = if { a } else { b } else { c };`.
+This is likely to be unclear if anyone writes it, but does not pose a syntactical issue, as `let PATTERN = if y { a } else { b };` should always be
+interpreted as `let Enum(x) = (if y { a } else { b });` (still a compile error as there no diverging block: `error[E0005]: refutable pattern in local binding: ...`)
+because the compiler won't interpret it as `let PATTERN = (if y { a }) else { b };` since `()` is not an enum.
+
+This can be overcome by making a raw if-else in the expression position a compile error and instead requiring that parentheses are inserted to disambiguate:
+`let PATTERN = (if { a } else { b }) else { c };`.
+
+Rust already provides us with such a restriction, and so the expression can be restricted to be a [`ExpressionWithoutBlock`][expressions].
+
+# Rationale and alternatives
+[rationale-and-alternatives]: #rationale-and-alternatives
+
+let-else attempts to be as consistent as possible to similar existing syntax.
+
+Fundamentally it is treated as a `let` statement, necessitating an assignment and the trailing semicolon.
+
+Pattern matching works identically to if-let, no new "negation" pattern matching rules are introduced.
+
+Operator precedence with `&&` in made to be like if-let, requiring that a case which is an error prior to this RFC be changed to be a slightly different error.
+This is for a possible extension for let-else similar to the (yet unimplemented) if-else-chains feature, as mentioned in [future-possibilities][] with more detail.
+Specifically, while the following example is an error today, by the default `&&` operator rules it would cause problems with if-let-chains like `&&` chaining:
+
+```rust
+let a = false;
+let b = false;
+
+// The RFC proposes boolean patterns with a lazy boolean operator (&& or ||)
+//  be made into a compile error, for potential future compatibility with if-let-chains.
+let true = a && b else {
+    return;
+};
+```
+
+The expression can be any [`ExpressionWithoutBlock`][expressions], in order to prevent `else {} else {}` confusion, as noted in [drawbacks][#drawbacks].
+
+The `else` must be followed by a block, as in `if {} else {}`. This else block must be diverging as the outer
+context cannot be guaranteed to continue soundly without assignment, and no alternate assignment syntax is provided.
+
+## Alternatives
```

Comment:
One alternative that I don't think is a good idea but is worth noting is some sort of "uplift" expression for matches. For example something like:

```rust
match thing {
  Happy(x) => let x, // Assigns x to outer scope.
  Sad(y) => return Err(format!("We were sad because of {}", y)),
  Tragic(z) => return Err(format!("We cried hard because of {}", z)),
}
```

The value of this approach over the proposed approach is that it allows handing multiple unexpected values more gracefully. For example the above would be awkward to write with the current RFC.

```rust
let Happy(x) = thing else {
  match thing {
    Happy(_) => unreachable!(),
    Sad(y) => return Err(format!("We were sad because of {}", y)),
    Tragic(z) => return Err(format!("We cried hard because of {}", z)),
  }
}
```

However I think that the "action at a distance" is really awkward and confusing. This could be possibly mitigated with something like the following, but I still don't think it is great.

```rust
let match thing {
  Happy(x) => let,
  // ...
}
```

So basically the current RFC extracts only a single pattern from the matching space and leaves it to the user to deal with the failure case with no support. This is not entirely different than `if let` but I think the context is worse because in the failed case of `if let` you are just left where you were before, in this case you have to diverge and the object you just matched against is likely critical to deciding how to diverge.

However I don't see this as a major issue. I think that in this case falling back to the previous `let (x) = match val { Happy(x) => x, /* ... */ }` scenario is reasonable. Overall I would love to see this RFC accepted.

Reply:
It's a bit silly to bikeshed a feature that isn't being proposed, but (IIRC) @scottmcm suggested `else match` for this, which might look like:
```rust
let Happy(x) = thing else match {
  Sad(y) => return Err(format!("We were sad because of {}", y)),
  Tragic(z) => return Err(format!("We cried hard because of {}", z)),
};
```
This can also be used in the else branch of an `if let`.

Reply:
That does seem quite nice and works well with this proposal. It still has some flaws (it isn't really consistent with any existing syntax) I think it would be an acceptable solution if it is deemed necessary later.

Reply:
Have noted this new alternative, and `else match` is already noted.